File: nozk_py/mock_test.py
"""Unit tests for MockMint (mint_mock.py) and MockRedeemer (redeem_mock.py).

BLS12-381 standard scheme: PK=G1, Sig=G2.
"""

import pytest

import nozk_library as gl
from bls12_381_crypto import CURVE_ORDER, serialize_g2_sol
from mint_mock import MockMint, MockMintError
from redeem_mock import MockRedeemer, NullifierState

# ==============================================================================
# FIXTURES
# ==============================================================================

_TEST_CHAIN_ID = 11155111
_TEST_CONTRACT = "0x00000000000000000000000000000000DeaDBeef"
_TEST_DEADLINE = 2**256 - 1


@pytest.fixture
def mint():
    return MockMint.from_sk(42)


@pytest.fixture
def keypair():
    return gl.generate_mint_keypair()


@pytest.fixture
def lifecycle(keypair):
    """Full lifecycle: derive secrets, blind, sign, unblind, generate redemption proof."""
    seed = b"mock_test_lifecycle_seed"
    secrets = gl.derive_token_secrets(seed, 0)
    blinded = gl.blind_token(secrets.spend_bls_pub, secrets.r)
    S_prime = gl.mint_blind_sign(blinded.B, keypair.sk)
    S = gl.unblind_signature(S_prime, secrets.r)
    recipient = "0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa"
    proof = gl.generate_redemption_proof(
        secrets.spend_chia_sk,
        secrets.spend_chia_pk,
        recipient,
        _TEST_CHAIN_ID,
        _TEST_CONTRACT,
        _TEST_DEADLINE,
    )
    return {
        "keypair": keypair,
        "secrets": secrets,
        "blinded": blinded,
        "S": S,
        "recipient": recipient,
        "proof": proof,
    }


# ==============================================================================
# MockMint CONSTRUCTORS
# ==============================================================================


def test_mock_mint_from_sk_valid():
    m = MockMint.from_sk(42)
    assert m.sk == 42


def test_mock_mint_from_sk_zero_raises():
    with pytest.raises(MockMintError):
        MockMint.from_sk(0)


def test_mock_mint_from_sk_too_large_raises():
    with pytest.raises(MockMintError):
        MockMint.from_sk(CURVE_ORDER)


def test_mock_mint_from_hex_with_prefix():
    m = MockMint.from_hex("0x2a")
    assert m.sk == 42


def test_mock_mint_from_hex_invalid_raises():
    with pytest.raises(MockMintError):
        MockMint.from_hex("not_hex")


# ==============================================================================
# MockMint SIGNING
# ==============================================================================


def test_mock_mint_sign_returns_g2_point(mint):
    secrets = gl.derive_token_secrets(b"sign_test", 0)
    blinded = gl.blind_token(secrets.spend_bls_pub, secrets.r)
    S_prime = mint.sign(blinded.B)
    coords = serialize_g2_sol(S_prime)
    assert len(coords) == 8
    assert any(c != 0 for c in coords)


def test_mock_mint_sign_and_serialize_returns_int_tuple(mint):
    secrets = gl.derive_token_secrets(b"serialize_test", 0)
    blinded = gl.blind_token(secrets.spend_bls_pub, secrets.r)
    coords = mint.sign_and_serialize(blinded.B)
    assert len(coords) == 8
    assert all(isinstance(c, int) for c in coords)


# ==============================================================================
# MockRedeemer — Reveal
# ==============================================================================


def test_reveal_valid_bls(lifecycle):
    redeemer = MockRedeemer.from_sk(lifecycle["keypair"].sk)
    secrets = lifecycle["secrets"]
    result = redeemer.reveal(secrets.spend_bls_pub, lifecycle["S"])
    assert result.success is True
    assert result.bls_pairing_ok is True
    nid = secrets.nullifier_id.hex()
    assert redeemer.get_state(nid) == NullifierState.REVEALED


def test_reveal_already_revealed(lifecycle):
    redeemer = MockRedeemer.from_sk(lifecycle["keypair"].sk)
    secrets = lifecycle["secrets"]
    redeemer.reveal(secrets.spend_bls_pub, lifecycle["S"])
    result = redeemer.reveal(secrets.spend_bls_pub, lifecycle["S"])
    assert result.success is False
    assert "already" in (result.reason or "").lower()


def test_reveal_wrong_mint_key(lifecycle):
    wrong_kp = gl.generate_mint_keypair()
    redeemer = MockRedeemer.from_sk(wrong_kp.sk)
    secrets = lifecycle["secrets"]
    result = redeemer.reveal(secrets.spend_bls_pub, lifecycle["S"])
    assert result.success is False
    assert result.bls_pairing_ok is False


# ==============================================================================
# MockRedeemer — Redeem (requires prior reveal)
# ==============================================================================


def test_redeem_after_reveal(lifecycle):
    redeemer = MockRedeemer.from_sk(lifecycle["keypair"].sk)
    secrets = lifecycle["secrets"]

    rev = redeemer.reveal(secrets.spend_bls_pub, lifecycle["S"])
    assert rev.success is True

    nid = secrets.nullifier_id.hex()
    result = redeemer.redeem(
        recipient=lifecycle["recipient"],
        sigma=lifecycle["proof"].sigma,
        spend_pk=lifecycle["proof"].spend_pk,
        nullifier_id=nid,
        chain_id=_TEST_CHAIN_ID,
        contract_address=_TEST_CONTRACT,
        deadline=_TEST_DEADLINE,
    )
    assert result.success is True
    assert result.bls_spend_ok is True
    assert redeemer.get_state(nid) == NullifierState.SPENT


def test_redeem_without_reveal_fails(lifecycle):
    redeemer = MockRedeemer.from_sk(lifecycle["keypair"].sk)
    nid = lifecycle["secrets"].nullifier_id.hex()
    result = redeemer.redeem(
        recipient=lifecycle["recipient"],
        sigma=lifecycle["proof"].sigma,
        spend_pk=lifecycle["proof"].spend_pk,
        nullifier_id=nid,
        chain_id=_TEST_CHAIN_ID,
        contract_address=_TEST_CONTRACT,
        deadline=_TEST_DEADLINE,
    )
    assert result.success is False
    assert "not revealed" in (result.reason or "").lower()


def test_double_spend(lifecycle):
    redeemer = MockRedeemer.from_sk(lifecycle["keypair"].sk)
    secrets = lifecycle["secrets"]
    nid = secrets.nullifier_id.hex()

    redeemer.reveal(secrets.spend_bls_pub, lifecycle["S"])

    r1 = redeemer.redeem(
        recipient=lifecycle["recipient"],
        sigma=lifecycle["proof"].sigma,
        spend_pk=lifecycle["proof"].spend_pk,
        nullifier_id=nid,
        chain_id=_TEST_CHAIN_ID,
        contract_address=_TEST_CONTRACT,
        deadline=_TEST_DEADLINE,
    )
    assert r1.success is True

    r2 = redeemer.redeem(
        recipient=lifecycle["recipient"],
        sigma=lifecycle["proof"].sigma,
        spend_pk=lifecycle["proof"].spend_pk,
        nullifier_id=nid,
        chain_id=_TEST_CHAIN_ID,
        contract_address=_TEST_CONTRACT,
        deadline=_TEST_DEADLINE,
    )
    assert r2.success is False
    assert r2.nullifier_spent is True


def test_is_spent_and_reset(lifecycle):
    redeemer = MockRedeemer.from_sk(lifecycle["keypair"].sk)
    secrets = lifecycle["secrets"]
    nid = secrets.nullifier_id.hex()

    redeemer.reveal(secrets.spend_bls_pub, lifecycle["S"])
    result = redeemer.redeem(
        recipient=lifecycle["recipient"],
        sigma=lifecycle["proof"].sigma,
        spend_pk=lifecycle["proof"].spend_pk,
        nullifier_id=nid,
        chain_id=_TEST_CHAIN_ID,
        contract_address=_TEST_CONTRACT,
        deadline=_TEST_DEADLINE,
    )
    assert result.success is True
    assert redeemer.is_spent(nid)

    redeemer.reset()
    assert not redeemer.is_spent(nid)

File: nozk_ts/mock.test.ts
/**
 * Mock infrastructure tests — mirrors Python mock_test.py.
 *
 * Tests MockMint (blind signing) and MockRedeemer (state machine
 * simulating reveal/redeem with BLS verification).
 */

import { describe, expect, it } from 'vitest';
import { CURVE_ORDER, serializeG2Sol } from './bls12-381-crypto.js';
import { MockMint, MockMintError } from './mint-mock.js';
import {
    blindToken,
    deriveTokenSecrets,
    generateRedemptionProof,
    getNullifierIdHex,
    mintBlindSign,
    unblindSignature,
} from './nozk-library.js';
import { MockRedeemer, NullifierState } from './redeem-mock.js';

// ==============================================================================
// SHARED CONSTANTS
// ==============================================================================

const TEST_CHAIN_ID = 11155111;
const TEST_CONTRACT = '0x00000000000000000000000000000000DeaDBeef';
const TEST_DEADLINE = (1n << 256n) - 1n;

// ==============================================================================
// HELPERS
// ==============================================================================

function setupLifecycle(keypairSk: bigint) {
    const seed = new TextEncoder().encode('mock_test_lifecycle_seed');
    const secrets = deriveTokenSecrets(seed, 0);
    const { Y, B } = blindToken(secrets.spendBlsPub, secrets.r);
    const S_prime = mintBlindSign(B, keypairSk);
    const S = unblindSignature(S_prime, secrets.r);
    const recipient = '0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa';
    const proof = generateRedemptionProof(
        secrets.spendBlsPriv,
        secrets.spendPubCompressed,
        recipient,
        TEST_CHAIN_ID,
        TEST_CONTRACT,
        TEST_DEADLINE,
    );
    return { secrets, Y, B, S, recipient, proof };
}

// ==============================================================================
// MockMint constructors
// ==============================================================================

describe('MockMint constructors', () => {
    it('fromSk with valid scalar', () => {
        const mint = MockMint.fromSk(42n);
        expect(mint.sk).toBe(42n);
    });

    it('fromSk with zero throws', () => {
        expect(() => MockMint.fromSk(0n)).toThrow(MockMintError);
    });

    it('fromSk with CURVE_ORDER throws', () => {
        expect(() => MockMint.fromSk(CURVE_ORDER)).toThrow(MockMintError);
    });

    it('fromHex with 0x prefix', () => {
        const mint = MockMint.fromHex('0x2a');
        expect(mint.sk).toBe(42n);
    });

    it('fromHex with invalid string throws', () => {
        expect(() => MockMint.fromHex('not_hex')).toThrow(MockMintError);
    });
});

// ==============================================================================
// MockMint signing
// ==============================================================================

describe('MockMint signing', () => {
    it('sign returns G2 point', () => {
        const mint = MockMint.fromSk(42n);
        const seed = new TextEncoder().encode('sign_test');
        const secrets = deriveTokenSecrets(seed, 0);
        const { B } = blindToken(secrets.spendBlsPub, secrets.r);
        const sPrime = mint.sign(B);
        const coords = serializeG2Sol(sPrime);
        expect(coords.some((c) => c !== 0n)).toBe(true);
    });

    it('signAndSerialize returns 8 bigints', () => {
        const mint = MockMint.fromSk(42n);
        const seed = new TextEncoder().encode('serialize_test');
        const secrets = deriveTokenSecrets(seed, 0);
        const { B } = blindToken(secrets.spendBlsPub, secrets.r);
        const coords = mint.signAndSerialize(B);
        expect(coords.length).toBe(8);
        expect(coords.every((c) => typeof c === 'bigint')).toBe(true);
    });
});

// ==============================================================================
// MockRedeemer reveal
// ==============================================================================

describe('MockRedeemer reveal', () => {
    it('valid reveal succeeds', () => {
        const sk = 42n;
        const redeemer = MockRedeemer.fromSk(sk);
        const { secrets, S } = setupLifecycle(sk);

        const result = redeemer.reveal(secrets.spendBlsPub, S);
        expect(result.success).toBe(true);
        expect(result.blsPairingOk).toBe(true);
        expect(redeemer.isRevealed(result.nullifierId)).toBe(true);
    });

    it('already revealed fails', () => {
        const sk = 42n;
        const redeemer = MockRedeemer.fromSk(sk);
        const { secrets, S } = setupLifecycle(sk);

        const first = redeemer.reveal(secrets.spendBlsPub, S);
        expect(first.success).toBe(true);

        const second = redeemer.reveal(secrets.spendBlsPub, S);
        expect(second.success).toBe(false);
        expect(second.reason).toContain('already');
    });

    it('wrong mint key fails BLS pairing', () => {
        const realSk = 42n;
        const wrongSk = 999n;
        const redeemer = MockRedeemer.fromSk(wrongSk); // wrong key
        const { secrets, S } = setupLifecycle(realSk); // signed with real key

        const result = redeemer.reveal(secrets.spendBlsPub, S);
        expect(result.success).toBe(false);
        expect(result.blsPairingOk).toBe(false);
    });
});

// ==============================================================================
// MockRedeemer redeem
// ==============================================================================

describe('MockRedeemer redeem', () => {
    it('redeem after reveal succeeds', () => {
        const sk = 42n;
        const redeemer = MockRedeemer.fromSk(sk);
        const { secrets, S, recipient, proof } = setupLifecycle(sk);

        redeemer.reveal(secrets.spendBlsPub, S);

        const nid = getNullifierIdHex(secrets);
        const result = redeemer.redeem({
            recipient,
            sigma: proof.sigma,
            spendPubCompressed: proof.spendPubCompressed,
            nullifierId: nid,
            chainId: TEST_CHAIN_ID,
            contractAddress: TEST_CONTRACT,
            deadline: TEST_DEADLINE,
        });
        expect(result.success).toBe(true);
        expect(result.blsSpendOk).toBe(true);
        expect(redeemer.isSpent(nid)).toBe(true);
    });

    it('redeem without reveal fails', () => {
        const sk = 42n;
        const redeemer = MockRedeemer.fromSk(sk);
        const { secrets, recipient, proof } = setupLifecycle(sk);

        const nid = getNullifierIdHex(secrets);
        const result = redeemer.redeem({
            recipient,
            sigma: proof.sigma,
            spendPubCompressed: proof.spendPubCompressed,
            nullifierId: nid,
            chainId: TEST_CHAIN_ID,
            contractAddress: TEST_CONTRACT,
            deadline: TEST_DEADLINE,
        });
        expect(result.success).toBe(false);
        expect(result.reason).toContain('not revealed');
    });

    it('double spend fails', () => {
        const sk = 42n;
        const redeemer = MockRedeemer.fromSk(sk);
        const { secrets, S, recipient, proof } = setupLifecycle(sk);

        redeemer.reveal(secrets.spendBlsPub, S);

        const nid = getNullifierIdHex(secrets);
        const params = {
            recipient,
            sigma: proof.sigma,
            spendPubCompressed: proof.spendPubCompressed,
            nullifierId: nid,
            chainId: TEST_CHAIN_ID,
            contractAddress: TEST_CONTRACT,
            deadline: TEST_DEADLINE,
        };

        const first = redeemer.redeem(params);
        expect(first.success).toBe(true);

        const second = redeemer.redeem(params);
        expect(second.success).toBe(false);
        expect(second.nullifierSpent).toBe(true);
    });

    it('isSpent + reset', () => {
        const sk = 42n;
        const redeemer = MockRedeemer.fromSk(sk);
        const { secrets, S, recipient, proof } = setupLifecycle(sk);

        redeemer.reveal(secrets.spendBlsPub, S);
        const nid = getNullifierIdHex(secrets);
        redeemer.redeem({
            recipient,
            sigma: proof.sigma,
            spendPubCompressed: proof.spendPubCompressed,
            nullifierId: nid,
            chainId: TEST_CHAIN_ID,
            contractAddress: TEST_CONTRACT,
            deadline: TEST_DEADLINE,
        });

        expect(redeemer.isSpent(nid)).toBe(true);
        expect(redeemer.getState(nid)).toBe(NullifierState.SPENT);

        redeemer.reset();
        expect(redeemer.isSpent(nid)).toBe(false);
        expect(redeemer.getState(nid)).toBe(NullifierState.UNREVEALED);
    });
});
